File: backend/app/schemas/packages.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class PackageUpdatePublic(BaseModel):
    name: str
    current_version: str | None = None
    candidate_version: str | None = None
    is_security: bool = False
    origin: str | None = None
    architecture: str | None = None


class PackageStatePublic(BaseModel):
    manager: str
    updates: list[PackageUpdatePublic] = []
    security_count: int = 0
    reboot_required: bool = False
    reboot_required_by: list[str] = []
    last_refreshed_iso: str | None = None


class PackageRunPublic(BaseModel):
    id: UUID
    kind: Literal["refresh", "upgrade"]
    state: Literal["running", "succeeded", "failed", "interrupted"]
    packages: str | None = None
    started_at: datetime
    finished_at: datetime | None = None
    output: str | None = None
    error: str | None = None

    model_config = {"from_attributes": True}


class PackageUpgradeRequest(BaseModel):
    """Empty `packages` means everything upgradable — the common case, and
    the one the UI's main button uses."""

    packages: list[str] = Field(default_factory=list, max_length=200)
    security_only: bool = False

File: frontend/src/lib/packages.ts
import { api, readErrorMessage } from "@/lib/api";

export interface PackageUpdate {
  name: string;
  current_version: string | null;
  candidate_version: string | null;
  is_security: boolean;
  origin: string | null;
  architecture: string | null;
}

export interface PackageState {
  manager: string;
  updates: PackageUpdate[];
  security_count: number;
  reboot_required: boolean;
  reboot_required_by: string[];
  last_refreshed_iso: string | null;
}

export type PackageRunState =
  | "running"
  | "succeeded"
  | "failed"
  | "interrupted";

export interface PackageRun {
  id: string;
  kind: "refresh" | "upgrade";
  state: PackageRunState;
  packages: string | null;
  started_at: string;
  finished_at: string | null;
  output: string | null;
  error: string | null;
}

export async function getPackages(deviceId: string): Promise<PackageState> {
  return api
    .get(`devices/${deviceId}/packages`, { timeout: 90_000 })
    .json<PackageState>();
}

export async function listPackageRuns(deviceId: string): Promise<PackageRun[]> {
  return api.get(`devices/${deviceId}/package-runs`).json<PackageRun[]>();
}

export async function refreshPackages(deviceId: string): Promise<PackageRun> {
  try {
    return await api
      .post(`devices/${deviceId}/packages/refresh`)
      .json<PackageRun>();
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}

export async function upgradePackages(
  deviceId: string,
  packages: string[] = [],
  securityOnly = false,
): Promise<PackageRun> {
  try {
    return await api
      .post(`devices/${deviceId}/packages/upgrade`, {
        json: { packages, security_only: securityOnly },
      })
      .json<PackageRun>();
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}
